// @elastic/eui/lib/components/series_chart/status-text.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _text = require("../text");

var _icon = require("../icon");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function StatusText(_ref) {
  var width = _ref.width,
      height = _ref.height,
      text = _ref.text;
  return _react.default.createElement("div", {
    className: "euixychart-error-nodata",
    style: {
      width: width,
      height: height
    }
  }, _react.default.createElement("div", {
    style: {
      position: 'absolute',
      top: '50%',
      left: '50%',
      transform: "translate(-50%, -50%)",
      color: 'rgb(53, 129, 255)'
    }
  }, _react.default.createElement("div", {
    className: "euiToastHeader--withBody"
  }, _react.default.createElement(_icon.EuiIcon, {
    className: "euiToastHeader__icon",
    type: "visualizeApp",
    color: "subdued",
    size: "m",
    "aria-hidden": "true"
  }), _react.default.createElement("span", {
    className: "euiToastHeader__title"
  }, "Graph not avaliable")), text && _react.default.createElement(_text.EuiText, {
    size: "s"
  }, text)));
}

StatusText.propTypes = {
  text: _propTypes.default.string
};
var _default = StatusText;
exports.default = _default;
StatusText.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "StatusText",
  "props": {
    "text": {
      "type": {
        "name": "string"
      },
      "required": false,
      "description": ""
    }
  }
};
module.exports = exports.default;
