// @elastic/eui/es/components/form/range/range_highlight.js
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
export var EuiRangeHighlight = function EuiRangeHighlight(_ref) {
  var hasFocus = _ref.hasFocus,
      showTicks = _ref.showTicks,
      lowerValue = _ref.lowerValue,
      upperValue = _ref.upperValue,
      max = _ref.max,
      min = _ref.min;
  // Calculate the width the range based on value
  // const rangeWidth = (value - min) / (max - min);
  var leftPosition = (lowerValue - min) / (max - min);
  var rangeWidth = (upperValue - lowerValue) / (max - min);
  var rangeWidthStyle = {
    marginLeft: "".concat(leftPosition * 100, "%"),
    width: "".concat(rangeWidth * 100, "%")
  };
  var classes = classNames('euiRangeHighlight', {
    'euiRangeHighlight--hasTicks': showTicks
  });
  var progressClasses = classNames('euiRangeHighlight__progress', {
    'euiRangeHighlight__progress--hasFocus': hasFocus
  });
  return React.createElement("div", {
    className: classes
  }, React.createElement("div", {
    className: progressClasses,
    style: rangeWidthStyle
  }));
};
EuiRangeHighlight.propTypes = {
  hasFocus: PropTypes.bool,
  showTicks: PropTypes.bool,
  lowerValue: PropTypes.number.isRequired,
  upperValue: PropTypes.number.isRequired,
  max: PropTypes.number.isRequired,
  min: PropTypes.number.isRequired
};
EuiRangeHighlight.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "EuiRangeHighlight",
  "props": {
    "hasFocus": {
      "type": {
        "name": "bool"
      },
      "required": false,
      "description": ""
    },
    "showTicks": {
      "type": {
        "name": "bool"
      },
      "required": false,
      "description": ""
    },
    "lowerValue": {
      "type": {
        "name": "number"
      },
      "required": true,
      "description": ""
    },
    "upperValue": {
      "type": {
        "name": "number"
      },
      "required": true,
      "description": ""
    },
    "max": {
      "type": {
        "name": "number"
      },
      "required": true,
      "description": ""
    },
    "min": {
      "type": {
        "name": "number"
      },
      "required": true,
      "description": ""
    }
  }
};
