// @elastic/eui/lib/services/color/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "isColorDark", {
  enumerable: true,
  get: function get() {
    return _is_color_dark.isColorDark;
  }
});
Object.defineProperty(exports, "hexToRgb", {
  enumerable: true,
  get: function get() {
    return _hex_to_rgb.hexToRgb;
  }
});
Object.defineProperty(exports, "rgbToHex", {
  enumerable: true,
  get: function get() {
    return _rgb_to_hex.rgbToHex;
  }
});
Object.defineProperty(exports, "calculateContrast", {
  enumerable: true,
  get: function get() {
    return _luminance_and_contrast.calculateContrast;
  }
});
Object.defineProperty(exports, "calculateLuminance", {
  enumerable: true,
  get: function get() {
    return _luminance_and_contrast.calculateLuminance;
  }
});
Object.defineProperty(exports, "VISUALIZATION_COLORS", {
  enumerable: true,
  get: function get() {
    return _visualization_colors.VISUALIZATION_COLORS;
  }
});
Object.defineProperty(exports, "DEFAULT_VISUALIZATION_COLOR", {
  enumerable: true,
  get: function get() {
    return _visualization_colors.DEFAULT_VISUALIZATION_COLOR;
  }
});
Object.defineProperty(exports, "colorPalette", {
  enumerable: true,
  get: function get() {
    return _color_palette.colorPalette;
  }
});
Object.defineProperty(exports, "palettes", {
  enumerable: true,
  get: function get() {
    return _eui_palettes.palettes;
  }
});

var _is_color_dark = require("./is_color_dark");

var _hex_to_rgb = require("./hex_to_rgb");

var _rgb_to_hex = require("./rgb_to_hex");

var _luminance_and_contrast = require("./luminance_and_contrast");

var _visualization_colors = require("./visualization_colors");

var _color_palette = require("./color_palette");

var _eui_palettes = require("./eui_palettes");

require("./color_types");
