// @elastic/eui/lib/services/key_codes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.keyCodes = exports.RIGHT = exports.LEFT = exports.UP = exports.DOWN = exports.BACKSPACE = exports.TAB = exports.ESCAPE = exports.SPACE = exports.ENTER = void 0;
var ENTER = 13;
exports.ENTER = ENTER;
var SPACE = 32;
exports.SPACE = SPACE;
var ESCAPE = 27;
exports.ESCAPE = ESCAPE;
var TAB = 9;
exports.TAB = TAB;
var BACKSPACE = 8; // Arrow keys

exports.BACKSPACE = BACKSPACE;
var DOWN = 40;
exports.DOWN = DOWN;
var UP = 38;
exports.UP = UP;
var LEFT = 37;
exports.LEFT = LEFT;
var RIGHT = 39;
exports.RIGHT = RIGHT;
var keyCodes;
exports.keyCodes = keyCodes;

(function (keyCodes) {
  keyCodes[keyCodes["ENTER"] = 13] = "ENTER";
  keyCodes[keyCodes["SPACE"] = 32] = "SPACE";
  keyCodes[keyCodes["ESCAPE"] = 27] = "ESCAPE";
  keyCodes[keyCodes["TAB"] = 9] = "TAB";
  keyCodes[keyCodes["BACKSPACE"] = 8] = "BACKSPACE";
  keyCodes[keyCodes["DOWN"] = 40] = "DOWN";
  keyCodes[keyCodes["UP"] = 38] = "UP";
  keyCodes[keyCodes["LEFT"] = 37] = "LEFT";
  keyCodes[keyCodes["RIGHT"] = 39] = "RIGHT";
})(keyCodes || (exports.keyCodes = keyCodes = {}));
