// @elastic/eui/lib/components/date_picker/super_date_picker/quick_select_popover/quick_select.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EuiQuickSelect = void 0;

require("core-js/modules/es7.symbol.async-iterator");

require("core-js/modules/es6.symbol");

require("core-js/modules/es6.object.set-prototype-of");

require("core-js/modules/web.dom.iterable");

require("core-js/modules/es6.array.iterator");

require("core-js/modules/es6.object.keys");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireWildcard(require("react"));

var _moment = _interopRequireDefault(require("moment"));

var _datemath = _interopRequireDefault(require("@elastic/datemath"));

var _button = require("../../../button");

var _flex = require("../../../flex");

var _title = require("../../../title");

var _spacer = require("../../../spacer");

var _form = require("../../../form");

var _tool_tip = require("../../../tool_tip");

var _horizontal_rule = require("../../../horizontal_rule");

var _time_units = require("../time_units");

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var LAST = 'last';
var NEXT = 'next';
var timeTenseOptions = [{
  value: LAST,
  text: 'Last'
}, {
  value: NEXT,
  text: 'Next'
}];
var timeUnitsOptions = Object.keys(_time_units.timeUnits).map(function (key) {
  return {
    value: key,
    text: "".concat(_time_units.timeUnits[key], "s")
  };
});

var EuiQuickSelect =
/*#__PURE__*/
function (_Component) {
  _inherits(EuiQuickSelect, _Component);

  function EuiQuickSelect(props) {
    var _this;

    _classCallCheck(this, EuiQuickSelect);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(EuiQuickSelect).call(this, props));

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "onTimeTenseChange", function (evt) {
      _this.setState({
        timeTense: evt.target.value
      });
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "onTimeValueChange", function (evt) {
      var sanitizedValue = parseInt(evt.target.value, 10);

      _this.setState({
        timeValue: isNaN(sanitizedValue) ? '' : sanitizedValue
      });
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "onTimeUnitsChange", function (evt) {
      _this.setState({
        timeUnits: evt.target.value
      });
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "applyQuickSelect", function () {
      var _this$state = _this.state,
          timeTense = _this$state.timeTense,
          timeValue = _this$state.timeValue,
          timeUnits = _this$state.timeUnits;

      if (timeTense === NEXT) {
        _this.props.applyTime({
          start: 'now',
          end: "now+".concat(timeValue).concat(timeUnits),
          quickSelect: _objectSpread({}, _this.state)
        });

        return;
      }

      _this.props.applyTime({
        start: "now-".concat(timeValue).concat(timeUnits),
        end: 'now',
        quickSelect: _objectSpread({}, _this.state)
      });
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "getBounds", function () {
      var startMoment = _datemath.default.parse(_this.props.start);

      var endMoment = _datemath.default.parse(_this.props.end, {
        roundUp: true
      });

      return {
        min: startMoment && startMoment.isValid() ? startMoment : (0, _moment.default)().subtract(15, 'minute'),
        max: endMoment && endMoment.isValid() ? endMoment : (0, _moment.default)()
      };
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "stepForward", function () {
      var _this$getBounds = _this.getBounds(),
          min = _this$getBounds.min,
          max = _this$getBounds.max;

      var diff = max.diff(min);

      _this.props.applyTime({
        start: (0, _moment.default)(max).add(1, 'ms').toISOString(),
        end: (0, _moment.default)(max).add(diff + 1, 'ms').toISOString(),
        keepPopoverOpen: true
      });
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "stepBackward", function () {
      var _this$getBounds2 = _this.getBounds(),
          min = _this$getBounds2.min,
          max = _this$getBounds2.max;

      var diff = max.diff(min);

      _this.props.applyTime({
        start: (0, _moment.default)(min).subtract(diff + 1, 'ms').toISOString(),
        end: (0, _moment.default)(min).subtract(1, 'ms').toISOString(),
        keepPopoverOpen: true
      });
    });

    var _this$props$prevQuick = _this.props.prevQuickSelect,
        _timeTense = _this$props$prevQuick.timeTense,
        _timeValue = _this$props$prevQuick.timeValue,
        _timeUnits = _this$props$prevQuick.timeUnits;
    _this.state = {
      timeTense: _timeTense ? _timeTense : LAST,
      timeValue: _timeValue ? _timeValue : 15,
      timeUnits: _timeUnits ? _timeUnits : 'm'
    };
    return _this;
  }

  _createClass(EuiQuickSelect, [{
    key: "render",
    value: function render() {
      return _react.default.createElement(_react.Fragment, null, _react.default.createElement(_flex.EuiFlexGroup, {
        responsive: false,
        alignItems: "center",
        gutterSize: "s"
      }, _react.default.createElement(_flex.EuiFlexItem, null, _react.default.createElement(_title.EuiTitle, {
        size: "xxxs"
      }, _react.default.createElement("span", null, "Quick select"))), _react.default.createElement(_flex.EuiFlexItem, {
        grow: false
      }, _react.default.createElement(_tool_tip.EuiToolTip, {
        content: "Previous time window"
      }, _react.default.createElement(_button.EuiButtonIcon, {
        "aria-label": "Previous time window",
        iconType: "arrowLeft",
        onClick: this.stepBackward
      }))), _react.default.createElement(_flex.EuiFlexItem, {
        grow: false
      }, _react.default.createElement(_tool_tip.EuiToolTip, {
        content: "Next time window"
      }, _react.default.createElement(_button.EuiButtonIcon, {
        "aria-label": "Next time window",
        iconType: "arrowRight",
        onClick: this.stepForward
      })))), _react.default.createElement(_spacer.EuiSpacer, {
        size: "s"
      }), _react.default.createElement(_flex.EuiFlexGroup, {
        gutterSize: "s",
        responsive: false
      }, _react.default.createElement(_flex.EuiFlexItem, null, _react.default.createElement(_form.EuiFormRow, {
        compressed: true
      }, _react.default.createElement(_form.EuiSelect, {
        "aria-label": "Quick time tense",
        value: this.state.timeTense,
        options: timeTenseOptions,
        onChange: this.onTimeTenseChange
      }))), _react.default.createElement(_flex.EuiFlexItem, null, _react.default.createElement(_form.EuiFormRow, {
        compressed: true
      }, _react.default.createElement(_form.EuiFieldNumber, {
        "aria-label": "Quick time value",
        value: this.state.timeValue,
        onChange: this.onTimeValueChange
      }))), _react.default.createElement(_flex.EuiFlexItem, null, _react.default.createElement(_form.EuiFormRow, {
        compressed: true
      }, _react.default.createElement(_form.EuiSelect, {
        "aria-label": "Quick time units",
        value: this.state.timeUnits,
        options: timeUnitsOptions,
        onChange: this.onTimeUnitsChange
      }))), _react.default.createElement(_flex.EuiFlexItem, {
        grow: false
      }, _react.default.createElement(_form.EuiFormRow, null, _react.default.createElement(_button.EuiButton, {
        className: "euiQuickSelect__applyButton",
        size: "s",
        onClick: this.applyQuickSelect,
        disabled: this.state.timeValue === '' || this.state.timeValue <= 0
      }, "Apply")))), _react.default.createElement(_horizontal_rule.EuiHorizontalRule, {
        margin: "s"
      }));
    }
  }]);

  return EuiQuickSelect;
}(_react.Component);

exports.EuiQuickSelect = EuiQuickSelect;
EuiQuickSelect.propTypes = {
  applyTime: _propTypes.default.func.isRequired,
  start: _propTypes.default.string.isRequired,
  end: _propTypes.default.string.isRequired,
  prevQuickSelect: _propTypes.default.object
};
EuiQuickSelect.defaultProps = {
  prevQuickSelect: {}
};
EuiQuickSelect.__docgenInfo = {
  "description": "",
  "methods": [{
    "name": "onTimeTenseChange",
    "docblock": null,
    "modifiers": [],
    "params": [{
      "name": "evt",
      "type": null
    }],
    "returns": null
  }, {
    "name": "onTimeValueChange",
    "docblock": null,
    "modifiers": [],
    "params": [{
      "name": "evt",
      "type": null
    }],
    "returns": null
  }, {
    "name": "onTimeUnitsChange",
    "docblock": null,
    "modifiers": [],
    "params": [{
      "name": "evt",
      "type": null
    }],
    "returns": null
  }, {
    "name": "applyQuickSelect",
    "docblock": null,
    "modifiers": [],
    "params": [],
    "returns": null
  }, {
    "name": "getBounds",
    "docblock": null,
    "modifiers": [],
    "params": [],
    "returns": null
  }, {
    "name": "stepForward",
    "docblock": null,
    "modifiers": [],
    "params": [],
    "returns": null
  }, {
    "name": "stepBackward",
    "docblock": null,
    "modifiers": [],
    "params": [],
    "returns": null
  }],
  "displayName": "EuiQuickSelect",
  "props": {
    "prevQuickSelect": {
      "defaultValue": {
        "value": "{}",
        "computed": false
      },
      "type": {
        "name": "object"
      },
      "required": false,
      "description": ""
    },
    "applyTime": {
      "type": {
        "name": "func"
      },
      "required": true,
      "description": ""
    },
    "start": {
      "type": {
        "name": "string"
      },
      "required": true,
      "description": ""
    },
    "end": {
      "type": {
        "name": "string"
      },
      "required": true,
      "description": ""
    }
  }
};
