// @elastic/eui/es/components/series_chart/axis/default_axis.js
import "core-js/modules/es7.symbol.async-iterator";
import "core-js/modules/es6.symbol";
import "core-js/modules/es6.object.assign";
import "core-js/modules/web.dom.iterable";
import "core-js/modules/es6.array.iterator";
import "core-js/modules/es6.object.keys";
import "core-js/modules/es6.object.set-prototype-of";

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

import React, { Fragment, PureComponent } from 'react';
import PropTypes from 'prop-types';
import { EuiXAxis } from './x_axis';
import { EuiYAxis } from './y_axis';
import { EuiHorizontalGrid } from './horizontal_grid';
import { EuiVerticalGrid } from './vertical_grid';
import { ORIENTATION } from '../utils/chart_utils';
/**
 * The Default Axis component, with X and Y axis on the bottom and left position respectively,
 * and horiznontal or vertical grid depending on the orientation prop.
 */

export var EuiDefaultAxis =
/*#__PURE__*/
function (_PureComponent) {
  _inherits(EuiDefaultAxis, _PureComponent);

  function EuiDefaultAxis() {
    _classCallCheck(this, EuiDefaultAxis);

    return _possibleConstructorReturn(this, _getPrototypeOf(EuiDefaultAxis).apply(this, arguments));
  }

  _createClass(EuiDefaultAxis, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          showGridLines = _this$props.showGridLines,
          orientation = _this$props.orientation,
          xOnZero = _this$props.xOnZero,
          yOnZero = _this$props.yOnZero,
          rest = _objectWithoutProperties(_this$props, ["showGridLines", "orientation", "xOnZero", "yOnZero"]);

      return React.createElement(Fragment, null, showGridLines && orientation === ORIENTATION.VERTICAL && React.createElement(EuiHorizontalGrid, rest), showGridLines && orientation === ORIENTATION.HORIZONTAL && React.createElement(EuiVerticalGrid, rest), React.createElement(EuiXAxis, _extends({
        onZero: xOnZero
      }, rest)), React.createElement(EuiYAxis, _extends({
        onZero: yOnZero
      }, rest)));
    }
  }]);

  return EuiDefaultAxis;
}(PureComponent);
EuiDefaultAxis.displayName = 'EuiDefaultAxis';
EuiDefaultAxis.propTypes = {
  /** The orientation of the chart, used to determine the correct orientation of grids */
  orientation: PropTypes.string,

  /** Show/Hide the background grids */
  showGridLines: PropTypes.bool,

  /** Specify if the x axis lay on 0, otherwise lyed on min x */
  xOnZero: PropTypes.bool,

  /** Specify if the y axis lay on 0, otherwise layd on min y */
  yOnZero: PropTypes.bool
};
EuiDefaultAxis.defaultProps = {
  orientation: ORIENTATION.VERTICAL,
  showGridLines: true,
  xOnZero: false,
  yOnZero: false
};
EuiDefaultAxis.requiresSVG = true;
EuiDefaultAxis.__docgenInfo = {
  "description": "The Default Axis component, with X and Y axis on the bottom and left position respectively,\nand horiznontal or vertical grid depending on the orientation prop.",
  "methods": [],
  "displayName": "EuiDefaultAxis",
  "props": {
    "orientation": {
      "defaultValue": {
        "value": "ORIENTATION.VERTICAL",
        "computed": true
      },
      "type": {
        "name": "string"
      },
      "required": false,
      "description": "The orientation of the chart, used to determine the correct orientation of grids"
    },
    "showGridLines": {
      "defaultValue": {
        "value": "true",
        "computed": false
      },
      "type": {
        "name": "bool"
      },
      "required": false,
      "description": "Show/Hide the background grids"
    },
    "xOnZero": {
      "defaultValue": {
        "value": "false",
        "computed": false
      },
      "type": {
        "name": "bool"
      },
      "required": false,
      "description": "Specify if the x axis lay on 0, otherwise lyed on min x"
    },
    "yOnZero": {
      "defaultValue": {
        "value": "false",
        "computed": false
      },
      "type": {
        "name": "bool"
      },
      "required": false,
      "description": "Specify if the y axis lay on 0, otherwise layd on min y"
    }
  }
};
