// @elastic/eui/es/components/series_chart/axis/x_axis.js
import "core-js/modules/es7.symbol.async-iterator";
import "core-js/modules/es6.symbol";
import "core-js/modules/es6.object.assign";
import "core-js/modules/web.dom.iterable";
import "core-js/modules/es6.array.iterator";
import "core-js/modules/es6.object.keys";
import "core-js/modules/es6.object.set-prototype-of";

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { XAxis } from 'react-vis';
import { EuiSeriesChartAxisUtils } from '../utils/axis_utils';
var TITLE_POSITION = EuiSeriesChartAxisUtils.TITLE_POSITION,
    ORIENTATION = EuiSeriesChartAxisUtils.ORIENTATION;
export var EuiXAxis =
/*#__PURE__*/
function (_PureComponent) {
  _inherits(EuiXAxis, _PureComponent);

  function EuiXAxis() {
    _classCallCheck(this, EuiXAxis);

    return _possibleConstructorReturn(this, _getPrototypeOf(EuiXAxis).apply(this, arguments));
  }

  _createClass(EuiXAxis, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          title = _this$props.title,
          titlePosition = _this$props.titlePosition,
          orientation = _this$props.orientation,
          tickSize = _this$props.tickSize,
          tickLabelAngle = _this$props.tickLabelAngle,
          tickFormat = _this$props.tickFormat,
          tickValues = _this$props.tickValues,
          onZero = _this$props.onZero,
          rest = _objectWithoutProperties(_this$props, ["title", "titlePosition", "orientation", "tickSize", "tickLabelAngle", "tickFormat", "tickValues", "onZero"]);

      return React.createElement(XAxis, _extends({
        title: title,
        position: titlePosition,
        orientation: orientation,
        on0: onZero,
        tickSize: tickSize,
        tickFormat: tickFormat,
        tickValues: tickValues,
        tickLabelAngle: tickLabelAngle
      }, rest));
    }
  }]);

  return EuiXAxis;
}(PureComponent);
EuiXAxis.displayName = 'EuiXAxis';
EuiXAxis.propTypes = {
  /** The axis title */
  title: PropTypes.string,

  /** The axis title position */
  titlePosition: PropTypes.oneOf([TITLE_POSITION.START, TITLE_POSITION.MIDDLE, TITLE_POSITION.END]),

  /** The axis orientation */
  orientation: PropTypes.oneOf([ORIENTATION.TOP, ORIENTATION.BOTTOM]),

  /** Fix the axis at zero value */
  onZero: PropTypes.bool,

  /** An array of ticks values */
  ticks: PropTypes.array,

  /** The height of the ticks in pixels */
  tickSize: PropTypes.number,

  /** TODO */
  tickValues: PropTypes.array,

  /** A formatter function in the form of function(value, index, scale, tickTotal) */
  tickFormat: PropTypes.func,

  /** the rotation angle in degree of the tick label */
  tickLabelAngle: PropTypes.number
};
EuiXAxis.defaultProps = {
  onZero: false,
  titlePosition: TITLE_POSITION.MIDDLE,
  orientation: ORIENTATION.BOTTOM,
  tickSize: 0
};
EuiXAxis.requiresSVG = true;
EuiXAxis.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "EuiXAxis",
  "props": {
    "onZero": {
      "defaultValue": {
        "value": "false",
        "computed": false
      },
      "type": {
        "name": "bool"
      },
      "required": false,
      "description": "Fix the axis at zero value"
    },
    "titlePosition": {
      "defaultValue": {
        "value": "TITLE_POSITION.MIDDLE",
        "computed": true
      },
      "type": {
        "name": "enum",
        "value": [{
          "value": "TITLE_POSITION.START",
          "computed": true
        }, {
          "value": "TITLE_POSITION.MIDDLE",
          "computed": true
        }, {
          "value": "TITLE_POSITION.END",
          "computed": true
        }]
      },
      "required": false,
      "description": "The axis title position"
    },
    "orientation": {
      "defaultValue": {
        "value": "ORIENTATION.BOTTOM",
        "computed": true
      },
      "type": {
        "name": "enum",
        "value": [{
          "value": "ORIENTATION.TOP",
          "computed": true
        }, {
          "value": "ORIENTATION.BOTTOM",
          "computed": true
        }]
      },
      "required": false,
      "description": "The axis orientation"
    },
    "tickSize": {
      "defaultValue": {
        "value": "0",
        "computed": false
      },
      "type": {
        "name": "number"
      },
      "required": false,
      "description": "The height of the ticks in pixels"
    },
    "title": {
      "type": {
        "name": "string"
      },
      "required": false,
      "description": "The axis title"
    },
    "ticks": {
      "type": {
        "name": "array"
      },
      "required": false,
      "description": "An array of ticks values"
    },
    "tickValues": {
      "type": {
        "name": "array"
      },
      "required": false,
      "description": "TODO"
    },
    "tickFormat": {
      "type": {
        "name": "func"
      },
      "required": false,
      "description": "A formatter function in the form of function(value, index, scale, tickTotal)"
    },
    "tickLabelAngle": {
      "type": {
        "name": "number"
      },
      "required": false,
      "description": "the rotation angle in degree of the tick label"
    }
  }
};
