// @elastic/eui/es/components/series_chart/series/area_series.js
import "core-js/modules/es7.symbol.async-iterator";
import "core-js/modules/es6.symbol";
import "core-js/modules/es6.object.assign";
import "core-js/modules/es6.object.keys";
import "core-js/modules/es6.object.set-prototype-of";
import "core-js/modules/web.dom.iterable";
import "core-js/modules/es6.array.iterator";
import "core-js/modules/es7.object.values";
import "core-js/modules/es6.function.name";

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React from 'react';
import PropTypes from 'prop-types';
import { AreaSeries, AbstractSeries, LineSeries } from 'react-vis';
import { CURVE } from '../utils/chart_utils'; // TODO: needs to send a PR to react-vis for incorporate these changes into AreaSeries class for vertical
// area chart visualizations.
// class ExtendedAreaSeries extends AreaSeries {
//   _renderArea(data, x, y0, y, curve, getNull) {
//     const x0 = this._getAttr0Functor('x');
//     let area = d3Area();
//     if (curve !== null) {
//       if (typeof curve === 'string' && curves[curve]) {
//         area = area.curve(curves[curve]);
//       } else if (typeof curve === 'function') {
//         area = area.curve(curve);
//       }
//     }
//     console.log(Object.getPrototypeOf(this))
//     area = area.defined(getNull);
//     area = area
//       .x1(x)
//       .x0(x0) // this is required for displaying vertical area charts.
//       .y0(y0)
//       .y1(y);
//     return area(data);
//   }
// }

export var EuiAreaSeries =
/*#__PURE__*/
function (_AbstractSeries) {
  _inherits(EuiAreaSeries, _AbstractSeries);

  function EuiAreaSeries() {
    var _getPrototypeOf2;

    var _this;

    _classCallCheck(this, EuiAreaSeries);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(EuiAreaSeries)).call.apply(_getPrototypeOf2, [this].concat(args)));

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "state", {
      isMouseOverSeries: false
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_onSeriesMouseOver", function () {
      _this.setState(function () {
        return {
          isMouseOverSeries: true
        };
      });
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_onSeriesMouseOut", function () {
      _this.setState(function () {
        return {
          isMouseOverSeries: false
        };
      });
    });

    return _this;
  }

  _createClass(EuiAreaSeries, [{
    key: "render",
    value: function render() {
      var isMouseOverSeries = this.state.isMouseOverSeries;

      var _this$props = this.props,
          name = _this$props.name,
          data = _this$props.data,
          curve = _this$props.curve,
          color = _this$props.color,
          lineSize = _this$props.lineSize,
          onSeriesClick = _this$props.onSeriesClick,
          fillOpacity = _this$props.fillOpacity,
          rest = _objectWithoutProperties(_this$props, ["name", "data", "curve", "color", "lineSize", "onSeriesClick", "fillOpacity"]);

      return React.createElement(React.Fragment, null, React.createElement(LineSeries, _extends({}, rest, {
        key: "".concat(name, "-line"),
        curve: curve,
        data: data,
        opacity: 1,
        onSeriesClick: onSeriesClick,
        color: color,
        style: {
          pointerEvents: 'visiblestroke',
          strokeWidth: lineSize
        }
      })), React.createElement(AreaSeries, _extends({
        key: "".concat(name, "-area"),
        className: "euiAreaSeries",
        curve: curve,
        color: color,
        data: data,
        onSeriesClick: onSeriesClick,
        onSeriesMouseOver: this._onSeriesMouseOver,
        onSeriesMouseOut: this._onSeriesMouseOut,
        style: {
          cursor: isMouseOverSeries && onSeriesClick ? 'pointer' : 'default',
          opacity: fillOpacity
        }
      }, rest)));
    }
  }]);

  return EuiAreaSeries;
}(AbstractSeries);
EuiAreaSeries.displayName = 'EuiAreaSeries';
EuiAreaSeries.propTypes = {
  /** The name used to define the data in tooltips and legends */
  name: PropTypes.string.isRequired,

  /** Array<{x: string|number, y: string|number}> */
  data: PropTypes.arrayOf(PropTypes.shape({
    x: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    y: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
  })).isRequired,

  /** See eui_palettes.js or colorPalette service for recommended colors */
  color: PropTypes.string,
  curve: PropTypes.oneOf(Object.values(CURVE)),
  onSeriesClick: PropTypes.func,
  lineSize: PropTypes.number,
  fillOpacity: PropTypes.number
};
EuiAreaSeries.defaultProps = {
  curve: CURVE.LINEAR,
  lineSize: 1,
  fillOpacity: 1
};
EuiAreaSeries.__docgenInfo = {
  "description": "",
  "methods": [{
    "name": "_onSeriesMouseOver",
    "docblock": null,
    "modifiers": [],
    "params": [],
    "returns": null
  }, {
    "name": "_onSeriesMouseOut",
    "docblock": null,
    "modifiers": [],
    "params": [],
    "returns": null
  }],
  "displayName": "EuiAreaSeries",
  "props": {
    "curve": {
      "defaultValue": {
        "value": "CURVE.LINEAR",
        "computed": true
      },
      "type": {
        "name": "enum",
        "computed": true,
        "value": "Object.values(CURVE)"
      },
      "required": false,
      "description": ""
    },
    "lineSize": {
      "defaultValue": {
        "value": "1",
        "computed": false
      },
      "type": {
        "name": "number"
      },
      "required": false,
      "description": ""
    },
    "fillOpacity": {
      "defaultValue": {
        "value": "1",
        "computed": false
      },
      "type": {
        "name": "number"
      },
      "required": false,
      "description": ""
    },
    "name": {
      "type": {
        "name": "string"
      },
      "required": true,
      "description": "The name used to define the data in tooltips and legends"
    },
    "data": {
      "type": {
        "name": "arrayOf",
        "value": {
          "name": "shape",
          "value": {
            "x": {
              "name": "union",
              "value": [{
                "name": "string"
              }, {
                "name": "number"
              }],
              "required": false
            },
            "y": {
              "name": "union",
              "value": [{
                "name": "string"
              }, {
                "name": "number"
              }],
              "required": false
            }
          }
        }
      },
      "required": true,
      "description": "Array<{x: string|number, y: string|number}>"
    },
    "color": {
      "type": {
        "name": "string"
      },
      "required": false,
      "description": "See eui_palettes.js or colorPalette service for recommended colors"
    },
    "onSeriesClick": {
      "type": {
        "name": "func"
      },
      "required": false,
      "description": ""
    }
  }
};
