// @elastic/eui/es/components/series_chart/status-text.js
import React from 'react';
import PropTypes from 'prop-types';
import { EuiText } from '../text';
import { EuiIcon } from '../icon';

function StatusText(_ref) {
  var width = _ref.width,
      height = _ref.height,
      text = _ref.text;
  return React.createElement("div", {
    className: "euixychart-error-nodata",
    style: {
      width: width,
      height: height
    }
  }, React.createElement("div", {
    style: {
      position: 'absolute',
      top: '50%',
      left: '50%',
      transform: "translate(-50%, -50%)",
      color: 'rgb(53, 129, 255)'
    }
  }, React.createElement("div", {
    className: "euiToastHeader--withBody"
  }, React.createElement(EuiIcon, {
    className: "euiToastHeader__icon",
    type: "visualizeApp",
    color: "subdued",
    size: "m",
    "aria-hidden": "true"
  }), React.createElement("span", {
    className: "euiToastHeader__title"
  }, "Graph not avaliable")), text && React.createElement(EuiText, {
    size: "s"
  }, text)));
}

StatusText.propTypes = {
  text: PropTypes.string
};
export default StatusText;
StatusText.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "StatusText",
  "props": {
    "text": {
      "type": {
        "name": "string"
      },
      "required": false,
      "description": ""
    }
  }
};
