// @elastic/eui/lib/components/steps/steps_horizontal.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EuiStepsHorizontal = void 0;

require("core-js/modules/es6.object.assign");

require("core-js/modules/web.dom.iterable");

require("core-js/modules/es6.array.iterator");

require("core-js/modules/es6.object.keys");

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _step_horizontal = require("./step_horizontal");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function renderHorizontalSteps(steps) {
  return steps.map(function (step, index) {
    var children = step.children,
        className = step.className,
        disabled = step.disabled,
        isSelected = step.isSelected,
        onClick = step.onClick,
        rest = _objectWithoutProperties(step, ["children", "className", "disabled", "isSelected", "onClick"]);

    return _react.default.createElement(_step_horizontal.EuiStepHorizontal, _extends({
      className: className,
      key: index,
      step: index + 1,
      disabled: disabled,
      isSelected: isSelected,
      onClick: onClick
    }, rest), children);
  });
}

var EuiStepsHorizontal = function EuiStepsHorizontal(_ref) {
  var className = _ref.className,
      steps = _ref.steps,
      rest = _objectWithoutProperties(_ref, ["className", "steps"]);

  var classes = (0, _classnames.default)('euiStepsHorizontal', className);
  return _react.default.createElement("div", _extends({
    role: "tablist",
    className: classes
  }, rest), renderHorizontalSteps(steps));
};

exports.EuiStepsHorizontal = EuiStepsHorizontal;

var stepPropType = _propTypes.default.shape({
  isSelected: _propTypes.default.bool,
  disabled: _propTypes.default.bool,
  children: _propTypes.default.node
});

EuiStepsHorizontal.propTypes = {
  className: _propTypes.default.string,
  steps: _propTypes.default.arrayOf(stepPropType).isRequired
};
EuiStepsHorizontal.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "EuiStepsHorizontal",
  "props": {
    "className": {
      "type": {
        "name": "string"
      },
      "required": false,
      "description": ""
    },
    "steps": {
      "type": {
        "name": "arrayOf",
        "value": {
          "name": "custom",
          "raw": "stepPropType"
        }
      },
      "required": true,
      "description": ""
    }
  }
};
