// @elastic/eui/es/components/card/card_select.js
import "core-js/modules/es6.object.assign";
import "core-js/modules/web.dom.iterable";
import "core-js/modules/es6.array.iterator";
import "core-js/modules/es6.object.keys";

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { EuiButtonEmpty, COLORS as BUTTON_EMPTY_COLORS } from '../button/button_empty';
import { EuiI18n } from '../i18n';
export var EuiCardSelect = function EuiCardSelect(_ref) {
  var className = _ref.className,
      onClick = _ref.onClick,
      isSelected = _ref.isSelected,
      isDisabled = _ref.isDisabled,
      color = _ref.color,
      children = _ref.children,
      rest = _objectWithoutProperties(_ref, ["className", "onClick", "isSelected", "isDisabled", "color", "children"]);

  var child = euiCardSelectableText(isSelected, isDisabled, children);
  var selectClasses = classNames('euiCardSelect', "euiCardSelect--".concat(euiCardSelectableColor(color, isSelected)), className);
  return React.createElement(EuiButtonEmpty, _extends({
    className: selectClasses,
    color: color || 'text',
    size: "xs",
    onClick: onClick,
    isDisabled: isDisabled,
    iconType: isSelected ? 'check' : undefined
  }, rest), child);
};
export var EuiCardSelectProps = {
  className: PropTypes.string,

  /**
   * You must handle the click event in order to have a select button
   */
  onClick: PropTypes.func.isRequired,

  /**
   * Is in the selected state
   */
  isSelected: PropTypes.bool,
  isDisabled: PropTypes.bool,

  /**
   * Override the default color with one of the available colors from `EuiButtonEmpty`
   */
  color: PropTypes.oneOf(BUTTON_EMPTY_COLORS),

  /**
   * Override the content (text) of the button
   */
  children: PropTypes.node
};
EuiCardSelect.propTypes = EuiCardSelectProps;

function euiCardSelectableText(isSelected, isDisabled, children) {
  if (children) {
    return children;
  }

  var text;

  if (isSelected) {
    text = React.createElement(EuiI18n, {
      token: "euiCardSelect.selected",
      default: "Selected"
    });
  } else if (isDisabled) {
    text = React.createElement(EuiI18n, {
      token: "euiCardSelect.unavailable",
      default: "Unavailable"
    });
  } else {
    text = React.createElement(EuiI18n, {
      token: "euiCardSelect.select",
      default: "Select"
    });
  }

  return text;
}

export function euiCardSelectableColor(color, isSelected) {
  var calculatedColor;

  if (color) {
    calculatedColor = color;
  } else if (isSelected) {
    calculatedColor = 'success';
  } else {
    calculatedColor = 'text';
  }

  return calculatedColor;
}
EuiCardSelect.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "EuiCardSelect",
  "props": {
    "className": {
      "type": {
        "name": "string"
      },
      "required": false,
      "description": ""
    },
    "onClick": {
      "type": {
        "name": "func"
      },
      "required": true,
      "description": "You must handle the click event in order to have a select button"
    },
    "isSelected": {
      "type": {
        "name": "bool"
      },
      "required": false,
      "description": "Is in the selected state"
    },
    "isDisabled": {
      "type": {
        "name": "bool"
      },
      "required": false,
      "description": ""
    },
    "color": {
      "type": {
        "name": "enum",
        "computed": true,
        "value": "BUTTON_EMPTY_COLORS"
      },
      "required": false,
      "description": "Override the default color with one of the available colors from `EuiButtonEmpty`"
    },
    "children": {
      "type": {
        "name": "node"
      },
      "required": false,
      "description": "Override the content (text) of the button"
    }
  }
};
