// @elastic/eui/lib/components/header/header_alert/header_alert.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EuiHeaderAlert = void 0;

require("core-js/modules/es6.object.assign");

require("core-js/modules/web.dom.iterable");

require("core-js/modules/es6.array.iterator");

require("core-js/modules/es6.object.keys");

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _button = require("../../button");

var _flex = require("../../flex");

var _i18n = require("../../i18n");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var EuiHeaderAlert = function EuiHeaderAlert(_ref) {
  var action = _ref.action,
      className = _ref.className,
      date = _ref.date,
      text = _ref.text,
      title = _ref.title,
      rest = _objectWithoutProperties(_ref, ["action", "className", "date", "text", "title"]);

  var classes = (0, _classnames.default)('euiHeaderAlert', className);
  return _react.default.createElement(_i18n.EuiI18n, {
    token: "euiHeaderAlert.dismiss",
    default: "Dismiss"
  }, function (dismiss) {
    return _react.default.createElement("div", _extends({
      className: classes
    }, rest), _react.default.createElement(_button.EuiButtonIcon, {
      "aria-label": dismiss,
      iconType: "cross",
      size: "s",
      className: "euiHeaderAlert__dismiss"
    }), _react.default.createElement("div", {
      className: "euiHeaderAlert__title"
    }, title), _react.default.createElement("div", {
      className: "euiHeaderAlert__text"
    }, text), _react.default.createElement(_flex.EuiFlexGroup, {
      justifyContent: "spaceBetween"
    }, _react.default.createElement(_flex.EuiFlexItem, {
      grow: false
    }, _react.default.createElement("div", {
      className: "euiHeaderAlert__action euiLink"
    }, action)), _react.default.createElement(_flex.EuiFlexItem, {
      grow: false
    }, _react.default.createElement("div", {
      className: "euiHeaderAlert__date"
    }, date))));
  });
};

exports.EuiHeaderAlert = EuiHeaderAlert;
EuiHeaderAlert.propTypes = {
  action: _propTypes.default.node,
  className: _propTypes.default.string,
  date: _propTypes.default.node.isRequired,
  text: _propTypes.default.node,
  title: _propTypes.default.node.isRequired
};
EuiHeaderAlert.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "EuiHeaderAlert",
  "props": {
    "action": {
      "type": {
        "name": "node"
      },
      "required": false,
      "description": ""
    },
    "className": {
      "type": {
        "name": "string"
      },
      "required": false,
      "description": ""
    },
    "date": {
      "type": {
        "name": "node"
      },
      "required": true,
      "description": ""
    },
    "text": {
      "type": {
        "name": "node"
      },
      "required": false,
      "description": ""
    },
    "title": {
      "type": {
        "name": "node"
      },
      "required": true,
      "description": ""
    }
  }
};
